Subject: [PATCH] Make tokio version more forgiving
---
Index: banks-client/Cargo.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/banks-client/Cargo.toml b/banks-client/Cargo.toml
--- a/banks-client/Cargo.toml	(revision 0a3e52ba8b4f63c3675fa91fc89c4f54f69e5855)
+++ b/banks-client/Cargo.toml	(revision 38e15cc36081a83ef38bd7f53df5ec39eb439500)
@@ -17,7 +17,7 @@
 solana-sdk = { path = "../sdk", version = "=1.14.13" }
 tarpc = { version = "0.29.0", features = ["full"] }
 thiserror = "1.0"
-tokio = { version = "~1.14.1", features = ["full"] }
+tokio = { version = "1", features = ["full"] }
 tokio-serde = { version = "0.8", features = ["bincode"] }

 [dev-dependencies]
Index: client-test/Cargo.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/client-test/Cargo.toml b/client-test/Cargo.toml
--- a/client-test/Cargo.toml	(revision 0a3e52ba8b4f63c3675fa91fc89c4f54f69e5855)
+++ b/client-test/Cargo.toml	(revision 38e15cc36081a83ef38bd7f53df5ec39eb439500)
@@ -29,7 +29,7 @@
 solana-transaction-status = { path = "../transaction-status", version = "=1.14.13" }
 solana-version = { path = "../version", version = "=1.14.13" }
 systemstat = "0.1.11"
-tokio = { version = "~1.14.1", features = ["full"] }
+tokio = { version = "1", features = ["full"] }

 [dev-dependencies]
 solana-logger = { path = "../logger", version = "=1.14.13" }
Index: core/Cargo.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/core/Cargo.toml b/core/Cargo.toml
--- a/core/Cargo.toml	(revision 0a3e52ba8b4f63c3675fa91fc89c4f54f69e5855)
+++ b/core/Cargo.toml	(revision 38e15cc36081a83ef38bd7f53df5ec39eb439500)
@@ -63,7 +63,7 @@
 sys-info = "0.9.1"
 tempfile = "3.3.0"
 thiserror = "1.0"
-tokio = { version = "~1.14.1", features = ["full"] }
+tokio = { version = "1", features = ["full"] }
 trees = "0.4.2"

 [dev-dependencies]
Index: rpc-test/Cargo.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/rpc-test/Cargo.toml b/rpc-test/Cargo.toml
--- a/rpc-test/Cargo.toml	(revision 0a3e52ba8b4f63c3675fa91fc89c4f54f69e5855)
+++ b/rpc-test/Cargo.toml	(revision 38e15cc36081a83ef38bd7f53df5ec39eb439500)
@@ -26,7 +26,7 @@
 solana-streamer = { path = "../streamer", version = "=1.14.13" }
 solana-test-validator = { path = "../test-validator", version = "=1.14.13" }
 solana-transaction-status = { path = "../transaction-status", version = "=1.14.13" }
-tokio = { version = "~1.14.1", features = ["full"] }
+tokio = { version = "1", features = ["full"] }

 [dev-dependencies]
 solana-logger = { path = "../logger", version = "=1.14.13" }
Index: rpc/Cargo.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/rpc/Cargo.toml b/rpc/Cargo.toml
--- a/rpc/Cargo.toml	(revision 0a3e52ba8b4f63c3675fa91fc89c4f54f69e5855)
+++ b/rpc/Cargo.toml	(revision 38e15cc36081a83ef38bd7f53df5ec39eb439500)
@@ -53,7 +53,7 @@
 spl-token-2022 = { version = "=0.5.0", features = ["no-entrypoint"] }
 stream-cancel = "0.8.1"
 thiserror = "1.0"
-tokio = { version = "~1.14.1", features = ["full"] }
+tokio = { version = "1", features = ["full"] }
 tokio-util = { version = "0.6", features = ["codec", "compat"] }

 [dev-dependencies]
